sdk/core/azure-core/azure/core/pipeline/transport/_requests_basic.py

```diff
@@ -130,17 +130,25 @@ def __next__(self):
                 self.response.internal_response.close()
                 raise StopIteration()
             except (requests.exceptions.ChunkedEncodingError,
-                    requests.exceptions.ConnectionError):
+                    requests.exceptions.ConnectionError) as ex:
                 retry_total -= 1
                 if retry_total <= 0:
                     retry_active = False
                 else:
                     time.sleep(retry_interval)
                     headers = {'range': 'bytes=' + str(self.downloaded) + '-'}
```

Comment:
I'm starting to repeat myself at least 3 times over, but again: the original request saved in `self.request` can already have a range set, e.g. `range: bytes=1000-`. So what does this new range request do, exactly?

Also note that there are REST APIs that have special range requests, e.g. azure blob store GET used `x-ms-range, for more information please refer to the documentation here: https://docs.microsoft.com/en-us/rest/api/storageservices/specifying-the-range-header-for-blob-service-operations

So it would not even help to try parse this header of the original request here, because one can never be sure to catch all the special cases of APIs here.

Again, please do not retry, this cannot be done safely.

Reply:
I should also add that, in general `self.downloaded` is unreliable, because it is computed inconsistently, sometimes
as
```
self.downloaded += len(chunk)
```
and sometimes
```
self.downloaded += self._block_size
```

Both of them can be wrong, in general: requests configures urllib3 to do decoding based on the content-encoding header of the response (e.g. gunzip the data stream). It tries to read block_size bytes, then decodes them. However, for chunked http responses, the block_size is not actually used; instead, the chunks are iterated over.

So in essence, there is no way to reliably know how many of the original bytes were downloaded unless it's clear there is the special case of no http response chunking and no content decoding. This cannot be assumed at this point, as this is a general-purpose library.

Reply:
Hey @xiangyan99,

regarding the real word scenarios. We are using the azure blob storage in our machine learning applications that pull parquet data from the azure blob storage. Parquet files implement row group statistics, so a common optimisation is to only read the parts of the parquet file that have the row group statistics and then do pruning based on predicates. 

See https://github.com/mbr/simplekv/blob/master/simplekv/net/_azurestore_new.py#L194 for a file like interface to access the azure blob storage and http://peter-hoffmann.com/2020/understand-predicate-pushdown-on-rowgroup-level-in-parquet-with-pyarrow-and-python.html or an explanation about row group pruning.

Reply:
Thank you @jochen-ott-by @hoffmann , your information is really helpful.

Let me re-think about it. :)